### PR TITLE
Bug 1837564: pkg/terraform: add diagnostics errors for terraform apply operations

### PR DESCRIPTION
**Pull request by abhinavdahiya**
The terraform errors are tracked in a buffer. This buffers is then used to match against various
known conditions to understand the reasons for the errors.

This now allows the terraform apply to return specific errors in these cases instead of previous `failed to apply Terraform` constant
string message.

/cc @openshift/openshift-team-installer

### CODE REVIEW

File: pkg/terraform/diagnose.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+var conditions = []condition{{
+	match: regexp.MustCompile(`Error: Error creating Blob .*: Error copy/waiting`),
+
+	reason:  "Timeout",
+	message: `Copying the VHD to user environment was too slow, and timeout was reached for the success.`,
+}, {
+	match: regexp.MustCompile(`Error: Error Creating/Updating Subnet .*: network.SubnetsClient#CreateOrUpdate: .* Code="AnotherOperationInProgress" Message="Another operation on this or dependent resource is in progress`),
+
+	reason:  "AzureMultiOperationFailure",
+	message: `Creating Subnets failed because the Azure could not process multiple operations.`,
```
Comment (patrickdillon) [inline pkg/terraform/diagnose.go:45]: `the Azure` -> `Azure`

**Comment by abhinavdahiya** [comment]
@patrickdillon here is an example of the error logged by the installer https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/3535/pull-ci-openshift-installer-master-e2e-azure/541#1:build-log.txt%3A108

File: pkg/terraform/diagnose.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+// specific on the top, generic matches on the bottom.
+var conditions = []condition{{
+	match: regexp.MustCompile(`Error: Error creating Blob .*: Error copy/waiting`),
+
+	reason:  "Timeout",
+	message: `Copying the VHD to user environment was too slow, and timeout was reached for the success.`,
+}, {
+	match: regexp.MustCompile(`Error: Error Creating/Updating Subnet .*: network.SubnetsClient#CreateOrUpdate: .* Code="AnotherOperationInProgress" Message="Another operation on this or dependent resource is in progress`),
+
+	reason:  "AzureMultiOperationFailure",
```
Comment (wking) [inline pkg/terraform/diagnose.go:44]: Can we structure this so Azure conditions can be defined in a `pkg/terraform/azure` subpackage, so per-platform sub-teams can be assigned as owners for their platform's conditions?  There might be some generic conditions that apply to all platforms, but I'd expect most of the time, you'd want a whole bunch of platform-specific conditions from the platform the installer is using, plus maybe a handful of generic conditions.

File: pkg/terraform/diagnose.go (hunk 1 above)
Comment (wking) [reply pkg/terraform/diagnose.go]: Fixed by dabcc4d.

File: pkg/diagnostics/error.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+func (e *Err) Unwrap() error { return e.Orig }
+
+// Error returns a string representation of the Err. The returned value
+// is expected to be a single value.
+// The format of the error string returned is,
+// `error(<Reason>) from <Source>: <Message>: <Cause of Orig>`
+func (e *Err) Error() string {
+	buf := &bytes.Buffer{}
+	if len(e.Source) > 0 {
+		fmt.Fprintf(buf, "error(%s) from %q", e.Reason, e.Source)
```
Comment (wking) [inline pkg/diagnostics/error.go:44]: Do we need `%q` for `Source`, or can we use `%s` (here and in your later `Error from %q`)?  Both of those strings are compiled into the installer right?  I'd expect to use %q only if we were passing along some information slurped from the provider, where we weren't sure if there was going to be whitespace or other potentially confusing characters.

Comment (abhinavdahiya) [reply pkg/diagnostics/error.go]: fixed.

File: pkg/terraform/diagnose.go (hunk 2 above)
Comment (abhinavdahiya) [reply pkg/terraform/diagnose.go:44]: for i'm keeping this list in same file and we can split in into platforms in future work.

**Pull request merged**